src/hypothesis/strategies.py

```diff
@@ -695,6 +695,32 @@ def text(
 
 @cacheable
 @defines_strategy
+def strings_matching_regex(regex):
+    """Return strategy that generates strings that match given regex.
+
+    Regex can be either a string or compiled regex (through
+    :func:`re.compile`).
+
+    You can use regex flags (such as :const:`python:re.IGNORECASE`,
+    :const:`python:re.DOTALL` or :const:`python:re.UNICODE`) to control
+    generation. Flags can be passed either in compiled regex (specify flags in
+    call to :func:`re.compile`) or inside pattern with (?iLmsux) group.
+
+    Some tricky regular expressions are partly supported or not supported at
+    all. "^" and "$" do not affect generation. Positive lookahead/lookbehind
+    groups are considered normal groups. Negative lookahead/lookbehind groups
+    do not do anything. Ternary regex groups
+    ('(?(name)yes-pattern|no-pattern)') generate variants regardless of
+    presence of named group and rely on filtering to produce correct results
+    (therefore can be slower than straightforward generation).
+
+    """
```

Comment:
Hmm, it's actually not that big a deal - you can't *just* map `bytes` over the output as non-printable bytes tend to cause encoding issues, but it turns out that replacing the `u''.join`, `hunichr`, and `characters()` calls with byte-string equivalents removes unicode entirely and is a surprisingly small change.

I tried a few things that just dissolved under TypeErrors and UnicodeErrors, so I'd suggest pulling [the commit from my branch](https://github.com/Zac-HD/hypothesis-python/commits/text-matching-regex).  Feel free to rebase it away however you like; and note that it'll need some tests with bytestrings for full coverage.  Here's my dead-simple "does this crash":

    @given(text() | binary())
    def test_regex_strategy_does_not_crash(ex):
        # If the string we get is a valid regex pattern...
        try:
            re.compile(ex)
        except re.error:
            assume(False)
        # The only error we should be able to get is no matching examples (for eg 'a^b')
        try:
            strings_matching_regex(ex).example()
        except NoExamples:
            assume(False)